storage/memorybysubnet/peer_store.go

```diff
@@ -0,0 +1,595 @@
+// Package memorybysubnet implements the storage interface for a Chihaya
+// BitTorrent tracker keeping peer data in memory organized by a pre-configured
+// subnet.
+package memorybysubnet
+
+import (
+	"encoding/binary"
+	"errors"
+	"net"
+	"runtime"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/yaml.v2"
+
+	"github.com/chihaya/chihaya/bittorrent"
+	"github.com/chihaya/chihaya/storage"
+)
+
+func init() {
+	prometheus.MustRegister(promGCDurationMilliseconds)
+	prometheus.MustRegister(promInfohashesCount)
+
+	// Register the storage driver.
+	storage.RegisterDriver("memorybysubnet", driver{})
+}
+
+var promGCDurationMilliseconds = prometheus.NewHistogram(prometheus.HistogramOpts{
+	Name:    "chihaya_storage_gc_duration_milliseconds",
+	Help:    "The time it takes to perform storage garbage collection",
+	Buckets: prometheus.ExponentialBuckets(9.375, 2, 10),
+})
+
+var promInfohashesCount = prometheus.NewGauge(prometheus.GaugeOpts{
+	Name: "chihaya_storage_infohashes_count",
+	Help: "The number of Infohashes tracked",
+})
+
+// recordGCDuration records the duration of a GC sweep.
+func recordGCDuration(duration time.Duration) {
+	promGCDurationMilliseconds.Observe(float64(duration.Nanoseconds()) / float64(time.Millisecond))
+}
+
+// recordInfohashesDelta records a change in the number of Infohashes tracked.
+func recordInfohashesDelta(delta float64) {
+	promInfohashesCount.Add(delta)
+}
+
+type driver struct{}
+
+func (d driver) NewPeerStore(icfg interface{}) (storage.PeerStore, error) {
+	// Marshal the config back into bytes.
+	bytes, err := yaml.Marshal(icfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the bytes into the proper config type.
+	var cfg Config
+	err = yaml.Unmarshal(bytes, &cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return New(cfg)
+}
+
+// ErrInvalidGCInterval is returned for a GarbageCollectionInterval that is
+// less than or equal to zero.
+var ErrInvalidGCInterval = errors.New("invalid garbage collection interval")
+
+// Config holds the configuration of a memory PeerStore.
+type Config struct {
+	GarbageCollectionInterval   time.Duration `yaml:"gc_interval"`
+	PeerLifetime                time.Duration `yaml:"peer_lifetime"`
+	ShardCount                  int           `yaml:"shard_count"`
+	PreferredIPv4SubnetMaskBits int           `yaml:"preferred_ipv4_subnet_mask_bits"`
+	PreferredIPv6SubnetMaskBits int           `yaml:"preferred_ipv6_subnet_mask_bits"`
```

Comment:
This means the bit count rather than the actual bits, right? If yes: name is misleading.